### DISCUSSION TITLE
Interactively change color theme of bar

### DISCUSSION THREAD

**Discussion by basbebe** [FelixKratz/SketchyBar#159] [Q&A] [answered]
I would like to change the color scheme of my SketchyBar depending on the current system dark/light mode.

Currently I see three approaches in doing so:
- change every created item individually 
- query SketchyBar and change all results (maybe somehow handling highlight-colors and background-colors
- working with global variables and just restarting SketchyBar 

I already have a global variable that resembles system dark/light mode so the last option seams to be the most straight-forward and maintainable.

But maybe there is an option I'm not considering that maybe let's me avoid restarting SketchyBar?

**Comment by FelixKratz** [comment, marked as answer]
The most elegant solution to this would probably be to use the regex feature of SketchyBar. I implemented it for precisely such tasks:
The item `<name>` can be a regular expression inside of two '/': `/<regex>/`.
E.g.:
```bash
sketchybar --set '/.*/' icon.color=0xff000000
```
sets the icon color of all items matching the regex `.*` (so all items) to black. This works for any regex and all the other properties of course work as well. This way you can also set properties for any named groups if you employ the appropriate naming convention in your configuration file.

Also, you can subscribe to the `AppleInterfaceThemeChangedNotification` via a custom event that gets triggered instantaneously when the system theme changes:
```bash
sketchybar --add event theme_change AppleInterfaceThemeChangedNotification
```
and
```bash
sketchybar --subscribe <item that holds the theme change script> theme_change
```

**Reply by basbebe** [reply]
this!

thank you – exactly what I was looking for!

I currently am using [dark-mode-notify](https://github.com/bouk/dark-mode-notify) to run a script that takes care of all the themes (kitty, editor, fish, spacebar) and a global fish variable.  
Looks like I might not need it anymore then (?)

Do you know if there is also a NSDistributedNotification that triggers when input language changes?

I had an element in spacebar that checked `~/Library/Preferences/com.apple.HIToolbox.plist` periodically to show the current keyboard layout. Making this more responsive and removing the need for the loop would be nice.

I only found [this solution](https://apple.stackexchange.com/questions/414844/automatically-change-keyboard-shortcuts-when-keyboard-layout-changes/414904#414904) which would use a Launch Agent to watch for `~/Library/Preferences/com.apple.HIToolbox.plist` to change.

**Reply by FelixKratz** [reply]
Yes, there is a notification for input source change: `AppleSelectedInputSourcesChangedNotification` it is triggered when you change the keyboard language from the original menubar or via the ctrl - space shortcut.

**Reply by basbebe** [reply]
thanks!

**Reply by AlJohri** [reply]
@basbebe can you share your end to end solution for changing color theme with the system?

**Reply by basbebe** [reply]
@AlJohri it's several moving parts. Although I haven't looked at it for a while and most of it should be in my (slightly outdated…) [dotfiles](https://github.com/basbebe/dotfiles), here's the gist:

- Use `Library/LaunchAgents/ke.bou.dark-mode-notify.plist` to call `$HOME/.local/bin/change-theme` whenever [dark-mode-notify](https://github.com/bouk/dark-mode-notify) is triggered:

  ```xml
  <?xml version="1.0" encoding="UTF-8"?>
  <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"
  "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
  <dict>
      <key>Label</key>
      <string>ke.bou.dark-mode-notify</string>
      <key>KeepAlive</key>
      <true/>
      <key>StandardErrorPath</key>
      <string>/tmp/dark-mode-notify-stderr.log</string>
      <key>StandardOutPath</key>
      <string>/tmp/dark-mode-notify-stdout.log</string>
      <key>ProgramArguments</key>
      <array>
         <string>/usr/local/bin/dark-mode-notify</string>
         <string>/Users/<USERNAME>/.local/bin/change-theme</string>
      </array>
  </dict>
  </plist>
  ```

- `$HOME/.local/bin/change-theme` calls `$HOME/.local/bin/get-system-theme` to check the system's theme and toggle if no argument is given, also changing a Fish global variable and themes of Kitty, Kakoune (using Helix by now) and Sketchybar (currently non-functioning):

  ```sh
  #!/bin/dash
  
  # toggle / light / dark
  # when run without an argument will reset theme to system state.
  # invoked by dark-mode-notify when system dark mode changes
  # dark-mode-notify sets $DARKMODE to 1 or 0 depending on the system's state.
  
  arg=$1
  
  # get mode depending on system state (use $DARKMODE,
  # otherwise use function to query state)
  get_mode() {
    if [ -z "$DARKMODE" ]; then
      mode=$(~/.local/bin/get-system-theme)
    elif [ "$DARKMODE" -eq 1 ]; then
      mode=dark
    elif [ "$DARKMODE" -eq 0 ]; then
      mode=light
    fi
    echo "$mode"
  }
  
  set_kitty() {
    for instance in /tmp/kitty_listen*; do
      (/usr/local/bin/kitty @ --to unix:"${instance}" set-colors \
        --all --configured \
        ~/.config/kitty/themes/everforest-"${1}".conf &)
    done
  }
  
  set_kak() {
    for client in $(/usr/local/bin/kak -l | /usr/local/bin/rg -v '\('); do
      (echo "colorscheme gruvbox-${1}" | /usr/local/bin/kak -p "$client" &)
    done
  }
  
  set_bar() {
    /usr/local/bin/sketchybar --trigger theme_change MODE=$1
  }
  
  set_fish() {
    (/usr/bin/env fish -c "set -U OS_THEME $1" &)
  }
  
  set_mode() {
    # quit if called by dark-mode-notify and system theme and fish variable
    # already match
    [ -n "$DARKMODE" ] \
      && [ "$1" = "$(/usr/local/bin/fish -c 'echo $OS_THEME ')" ] \
      && exit
    case $1 in
      light)
        #only change system state if not being called by system state change
        # (dark-mode-notify)
        [ -z "$DARKMODE" ] \
          && (osascript -l JavaScript -e \
            "Application('System Events').appearancePreferences.darkMode = false" \
            >/dev/null &)
        mode=light
        ;;
      dark)
        #only change system state if not being called by system state change
        # (dark-mode-notify)
        [ -z "$DARKMODE" ] \
          && (osascript -l JavaScript -e \
            "Application('System Events').appearancePreferences.darkMode = true" \
            >/dev/null &)
        mode=dark
        ;;
    esac
  
    (set_kitty "$mode" &)
    (set_kak "$mode" &)
    (set_bar "$mode" &)
    (set_fish "$mode" &)
  }
  
  case $arg in
    light)
      set_mode light
      ;;
    dark)
      set_mode dark
      ;;
    toggle)
      case $(get_mode) in
        dark)
          set_mode light
          ;;
        light)
          set_mode dark
          ;;
      esac
      ;;
    *)
      set_mode "$(get_mode)"
      ;;
  esac
  
  exit
  ```

  `$HOME/.local/bin/get-system-theme`
  ```sh
  #!/bin/dash
  
  defaults read -g AppleInterfaceStyle >/dev/null 2>&1 \
    && echo dark \
    || echo light
  ```

- [skhd](https://github.com/koekeishiya/skhd) shortcut to toggle theme:

  ```sh
  […]
  
  hyper - d : ~/.local/bin/change-theme toggle
  
  […]
  ```

- `~/.config/fish/functions/fish_update_colors.fish` could do some things when the global variable changes but is currently unused:
  ```fish
  function fish_update_colors --on-variable OS_THEME
      # fish_set_ANSI_colors
      # set -x OS_THEME $OS_THEME
      # set -x BAT_THEME gruvbox-{$OS_THEME}
      # echo "theme changed: " $OS_THEME
  end
  ```

Hope this helps!
